File: DataStructures/Graphs/Djikstra.py
pofrom Priority_Queue import *
from Graph import *
import sys

def djikstra_shortest_path(graph, sourceVertex):

    min_heap = PriorityQueue(True)
    # Distance Map
    distance = {}
    # Parent Map, where is the vertex coming from
    parent = {}

    inifinite = sys.maxsize

    # Assing all the vertices maximum weigtage
    for vertex in graph.all_vertices.values():
        min_heap.add_task(inifinite, vertex)

    # Now put weight on sourceVertex as 0, since it points to itself
    min_heap.change_task_priority(0, sourceVertex)
    distance[sourceVertex] = 0
    parent[sourceVertex] = None

    while min_heap.is_empty() is False:
        # peek the topmost element of the minheap
        task = min_heap.peek_task()
        weight = min_heap.get_task_priority(task)
        # Extract min
        current = min_heap.pop_task()
        # Assign weight on the current vertex
        distance[current] = weight

        # Explore all the neighbors of current
        for edge in current.edges:
            # get the adjacent_vertex
            adjacent = get_other_vertex_for_edge(current, edge)

            # If minheap contains this adjacent_vertex
            if min_heap.contains_task(adjacent) is False:
                continue

            # Add distance of current vertex to the edge weight to get the distance of adjacent vertex from source vertex
            # when it goes through current vertex
            new_distance = distance[current] + edge.w
            if min_heap.get_task_priority(adjacent) > new_distance:
                # decrease the value in the minheap to the new value
                min_heap.change_task_priority(new_distance, adjacent)
                parent[adjacent] = current

    return distance




def get_other_vertex_for_edge(vertex, edge):
    if edge.u_vertex.key == vertex.key:
        return edge.v_vertex
    else:
        return edge.u_vertex


if __name__ == '__main__':
    graph = Graph(False)
    graph.add_edge(1,2,5)
    graph.add_edge(2,3,2)
    graph.add_edge(1,4,9)
    graph.add_edge(1,5,3)
    graph.add_edge(5,6,2)
    graph.add_edge(6,4,2)
    graph.add_edge(3,4,3)

    distance = djikstra_shortest_path(graph, graph.all_vertices[1])
    print(distance)

File: Java/Graph/Djikstra.java
import java.util.*;

class Dijkstra {

	public Map<Vertex<Integer>, Integer> shortestPath(Graph<Integer> graph, Vertex<Integer> sourceVertex) {

		// Min Heap to store vertices with cost
		BinaryMinHeap<Vertex<Integer>> minHeap = new BinaryMinHeap<>();

		// Distance Map from Source to all other vertices in the graph
		Map<Vertex<Integer>, Integer> distanceMap = new HashMap<>();

		// Parent Map to map which vertex is coming from which parent
		Map<Vertex<Integer>, Vertex<Integer>> parent = new HashMap<>();


		/**
			1. Create a minHeap with all vertices with their cost as INF
			2. Mark cost of source vertex as 0, and extract it from minHeap
			3. Put the vertex in distanceMap, with their distance (cost)
			2. Extract adjacent vertex of sourceVertex from minHeap with minimum cost
			3. 
		**/

		for(Vertex<Integer> vertex : graph.getAllVertices()) {
			minHeap.add(Integer.MAX_VALUE, vertex);
		}

		// Decrease the cost of current/source vertex as 0, as cost to itself is 0
		minHeap.decrease(sourceVertex, 0);

		// Put this in the distanceMap
		distanceMap.put(sourceVertex, 0);

		// Put the parent of source vertex as null
		parent.put(sourceVertex, null);

		// Iterate till heap is not empty
		while(!minHeap.isEmpty()) {
			// Extract the vertex with minimum cost, which has vertex and distance of that vertex from source vertex
			BinaryMinHeap<Vertex<Integer>>.Node minNode = minHeap.extractMinNode();

			Vertex<Integer> current = minNode.key;

			// update the shortest distance of current from the source vertex
			distanceMap.put(current, minNode.weight);

			// explore the neighbors of current vertex
			for(Edge<Integer> edge : current.getEdges()) {

				// Get the adjacent vertex from the edge
				Vertex<Integer> adjacentVertex = getVertexForEdge(current, edge);

				// If heap does not contain any adjacent vertex, that means vertex already has shortest distance from the source vertex
				if(!minHeap.containsData(adjacentVertex)) {
					continue;
				}


				// Add distance of current vertex weight to the adjacent vertex from source vertex
				int newDistance = distanceMap.get(current) + edge.getWeight();

				// If this total distance is less than current stored distance, then decrease the weight 
				if(minHeap.getWeight(adjacentVertex) > newDistance) {
					minHeap.decrease(adjacentVertex, newDistance);
					parent.put(adjacentVertex, current);
				}
			}
		}

		return distanceMap;
	}

	private Vertex<Integer> getVertexForEdge(Vertex<Integer> v, Edge<Integer> e) {
		return e.getUvertex().equals(v) ? e.getVvertex() : e.getUvertex();
	}

	public static void main(String[] args ){
		Graph<Integer> graph = new Graph<>(false);

		graph.addEdge(1, 2, 3);
        graph.addEdge(2, 3, 1);
        graph.addEdge(3, 1, 1);
        graph.addEdge(1, 4, 1);
        graph.addEdge(2, 4, 3);
        graph.addEdge(4, 5, 6);
        graph.addEdge(5, 6, 2);
        graph.addEdge(3, 5, 5);
        graph.addEdge(3, 6, 4);

        Dijkstra dsp = new Dijkstra();
        Vertex<Integer> sourceVertex = graph.getVertex(1);

        Map<Vertex<Integer>, Integer> distance = dsp.shortestPath(graph, sourceVertex);
        System.out.println(distance);
	}
}